### PR TITLE
My hw 04 (aleksei zhuchkov)

### PR DESCRIPTION
**Pull request by kelzerock**

### CODE REVIEW

File: HW04/aleksei_zhuchkov/lib/student.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  def initialize(student)
+    @name = student[:name].capitalize
+    @surname = student[:surname].capitalize
+    @storage_homework = []
+    @storage_notification_for_student = []
+    @storage_notification_for_mentor = []
+  end
+
+  def add_homework_to_student_storage(homework)
+    @storage_homework << homework
```
Comment (DashaSledopit) [inline HW04/aleksei_zhuchkov/lib/student.rb:16]: Please, correct similar issues everywhere
```suggestion
    storage_homework << homework
```

File: HW04/aleksei_zhuchkov/lib/student.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def notifications
```
Comment (DashaSledopit) [inline HW04/aleksei_zhuchkov/lib/student.rb:19]: I don't think any notification logic is the responsibility of the Student class

File: HW04/aleksei_zhuchkov/lib/student.rb (hunk 1 above)
Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/student.rb]: It done.

File: HW04/aleksei_zhuchkov/lib/student.rb (hunk 2 above)
Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/student.rb:19]: I hope, I was understand you right. I changed my code, pls check.

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    info_homework[:student].add_homework_to_student_storage(homework)
+    info_homework[:student].storage_notification_for_student << Notification.new('Add homework', homework.title)
+    homework
+  end
+
+  def subscribe_to!(student)
+    @storage_students << student
+  end
+
+  def notifications
```
[the remark this answers is not in the document]
Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/mentor.rb:26]: My code ready to check! Thx.

File: HW04/aleksei_zhuchkov/lib/util.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+def print_notification_for_students_and_mentor(name, surname, notifications, type_person)
```
Comment (DashaSledopit) [inline HW04/aleksei_zhuchkov/lib/util.rb:3]: Please, store the notifications' logic in the Notification class

Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/util.rb]: Hi, I have done it https://github.com/labs-ruby/homeworks-2022/pull/96/commits/2317202f0cf45739eda68d5119b5ad79bb38f418
please check.

File: HW04/aleksei_zhuchkov/lib/homework.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  attr_reader :title, :description, :student
+
+  def initialize(title, description, student)
+    @title = title
+    @description = description
+    @student = student
+    @storage_homework_answer = []
+    @homework_in_work = false
+    @homework_ready_for_check = false
+    @homework_accept_by_mentor = false
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/homework.rb:14]: Could you please add `status` for this

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+class Mentor
+  attr_accessor :storage_students
+  attr_reader :name, :surname
+
+  def initialize(mentor)
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/mentor.rb:7]:
```suggestion
  def initialize(name:, surname:)
```

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @name = mentor[:name].capitalize
+    @surname = mentor[:surname].capitalize
+    @storage_students = []
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/mentor.rb:10]:
```suggestion
    @students_list = []
```

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def add_homework(info_homework)
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/mentor.rb:13]:
```suggestion
  def add_homework(title:, description:, student:)
```

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    storage_students << student
+  end
+
+  def notifications
+    notification_message = Notification.new
+    notification_message.print_notification_for_students_and_mentor(name, surname, storage_students, 'mentor')
+  end
+
+  def mark_as_read!
+    storage_students.each { |student| student.storage_notification_for_mentor.clear }
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/mentor.rb:30]: This is a mentor class, mentor should not read students notifications, he should clear only his own notifications.

File: HW04/aleksei_zhuchkov/lib/notification.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def print_notification_for_student(name, surname, notifications)
+    puts "Notification for student (#{name} #{surname}):"
+    notifications.each(&:print_notification)
+  end
+
+  def print_notification
+    puts "Title: #{title}, discription: #{description}"
+  end
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/notification.rb:38]:
```suggestion
```

Just add a method for a student/mentor that will return their notifications (do not mix business logic and output to the console)

File: HW04/aleksei_zhuchkov/lib/student.rb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  attr_accessor :storage_homework, :storage_notification_for_student, :storage_notification_for_mentor
+  attr_reader :name, :surname
+
+  def initialize(student)
+    @name = student[:name].capitalize
+    @surname = student[:surname].capitalize
+    @storage_homework = []
+    @storage_notification_for_student = []
+    @storage_notification_for_mentor = []
+  end
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/student.rb:13]: same

File: HW04/aleksei_zhuchkov/lib/student.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  def homeworks
+    puts "Homework for student (#{name} #{surname})"
+    storage_homework.each do |el|
+      puts "Task: #{el.title}, discription: #{el.description}, \n\
+Is homework in work: #{el.homework_in_work}; \n\
+Is homework ready for check: #{el.homework_ready_for_check}, \n\
+Is homework access by mentor: #{el.homework_accept_by_mentor} \n"
+    end
+    puts '-----------------------------'
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/student.rb:37]:
```suggestion
```

just return all student homeworks

**Comment by kelzerock** [comment]
@B0RlS  I have realized all suggestion in my code. Ready to check.

File: HW04/aleksei_zhuchkov/lib/mentor.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+class Mentor
+  attr_accessor :students_list, :notifications_list
+  attr_reader :name, :surname
+
+  def initialize(name:, surname:)
+    @name = name.capitalize
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/mentor.rb:8]: You don't really need this method (capitalize), but you can leave it

File: HW04/aleksei_zhuchkov/lib/notification.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+class Notification
+  attr_reader :title, :description
+
+  def initialize(title = 'None', description = 'None')
+    @title = title
+    @description = description
+  end
+
+  def print_notifications(notifications)
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/notification.rb:11]: Again, I am very against any mixing of business logic and information output in the code.
If you want, you can simply return the information you want and display it in your app.rb however you like. 
In the second option, you can create a Printer class that will do this.
(I recommend the first option)

Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/notification.rb:11]: Again, I am very against any mixing of business logic and information output in the code.
If you want, you can simply return the information you want and display it in your app.rb however you like. 
In the second option, you can create a Printer class that will do this.
(I recommend the first option)

File: HW04/aleksei_zhuchkov/lib/mentor.rb (hunk 13 above)
Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/mentor.rb]: Hi, I've done it, you can check it in this commit https://github.com/labs-ruby/homeworks-2022/pull/96/commits/7a48c4d003421b1f3ba66d6d41e8a4137699535e

File: HW04/aleksei_zhuchkov/lib/notification.rb (hunk 14 above)
Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/notification.rb]: Hi, I have done it. My commit https://github.com/labs-ruby/homeworks-2022/pull/96/commits/60daa14b13bcf23ecb46a62da61e662f4206d2dd

Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/notification.rb]: Hi, I have done it. See my commit https://github.com/labs-ruby/homeworks-2022/pull/96/commits/60daa14b13bcf23ecb46a62da61e662f4206d2dd
tnx

File: HW04/aleksei_zhuchkov/lib/notification.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    @title = title
+    @description = description
+  end
+
+  def get_notifications(notifications)
+    return 'None notification' if notifications.empty?
+
+    result = []
+    notifications.each { |el| result << [el.title, el.description] }
+    puts result
```
Comment (B0RlS) [inline HW04/aleksei_zhuchkov/lib/notification.rb:16]:
```suggestion
   result
```

Comment (kelzerock) [reply HW04/aleksei_zhuchkov/lib/notification.rb]: I did it. Pls check https://github.com/labs-ruby/homeworks-2022/pull/96/commits/dae1efbc03eaff7f656807528fa5be337be720f5

**Pull request merged**